File: ios/Classes/modules/PoseEstimation.swift
//
//  PoseEstimation.swift
//  flutter_amanisdk_v2
//
//  Created by Deniz Can on 14.11.2022.
//

import AmaniSDK
import Flutter
import UIKit

@available(iOS 12.0, *)
class PoseEstimation {
  private let module = Amani.sharedInstance.poseEstimation()
  private var sdkView: SDKView!

  public func start(settings: PoseEstimationSettings, result: @escaping FlutterResult) {
    let vc = UIApplication.shared.windows.last?.rootViewController
    module.setScreenConfig(screenConfig: [
      .appBackgroundColor: settings.appBackgroundColor,
      .appFontColor: settings.appFontColor,
      .primaryButtonBackgroundColor: settings.primaryButtonBackgroundColor,
      .primaryButtonTextColor: settings.primaryButtonTextColor,
      .ovalBorderColor: settings.ovalBorderColor,
      .ovalBorderSuccessColor: settings.ovalBorderSuccessColor,
      .poseCount: settings.poseCount,
      .mainGuideVisibility: settings.mainGuideVisibility,
      .secondaryGuideVisibility: settings.secondaryGuideVisibility,
    ])

    module.setInfoMessages(infoMessages: [
      .faceIsOk: settings.faceIsOk,
      .notInArea: settings.notInArea,
      .faceTooSmall: settings.faceTooSmall,
      .faceTooBig: settings.faceTooBig,
      .completed: settings.completed,
      .turnRight: settings.turnRight,
      .turnLeft: settings.turnLeft,
      .turnUp: settings.turnUp,
      .turnDown: settings.turnDown,
      .lookStraight: settings.lookStraight,
      .errorMessage: settings.errorMessage,
      .next: settings.next,
      .tryAgain: settings.tryAgain,
      .errorTitle: settings.errorTitle,
      .informationScreenDesc1: settings.informationScreenDesc1,
      .informationScreenDesc2: settings.informationScreenDesc2,
      .informationScreenTitle: settings.informationScreenTitle,
      .wrongPose: settings.wrongPose,
      .descriptionHeader: settings.descriptionHeader,
    ])

    if
    let mainGuideUp = settings.mainGuideUp,
    let mainGuideDown = settings.mainGuideDown,
    let mainGuideLeft = settings.mainGuideLeft,
    let mainGuideRight = settings.mainGuideRight,
    let mainGuideStraight = settings.mainGuideStraight {
      module.setMainGuideImages(guideImages: [
        .mainGuideUp: UIImage(named: mainGuideUp)!,
        .mainGuideDown: UIImage(named: mainGuideDown)!,
        .mainGuideLeft: UIImage(named: mainGuideLeft)!,
        .mainGuideRight: UIImage(named: mainGuideRight)!,
        .mainGuideStraight: UIImage(named: mainGuideStraight)!,
      ])
    }
    
    if
    let secondaryGuideUp = settings.secondaryGuideUp,
    let secondaryGuideDown = settings.secondaryGuideDown,
    let secondaryGuideLeft = settings.secondaryGuideLeft,
    let secondaryGuideRight = settings.secondaryGuideRight
    {
      module.setSecondaryGuideImages(guideImages: [
        .secondaryGuideUp: UIImage(named: secondaryGuideUp)!,
        .secondaryGuideDown: UIImage(named: secondaryGuideDown)!,
        .secondaryGuideLeft: UIImage(named: secondaryGuideLeft)!,
        .secondaryGuideRight: UIImage(named: secondaryGuideRight)!,
      ])
    }

    module.setManualCropTimeout(Timeout: settings.manualCropTimeout)

    do {
      let moduleView = try module.start { image in
        let data = image.pngData()
        result(FlutterStandardTypedData(bytes: data!))
        
        DispatchQueue.main.async {
          self.sdkView.removeFromSuperview()
        }
      }
      
      sdkView = SDKView(sdkView: moduleView!)
      sdkView.start(on: vc!)
      sdkView.setupBackButton(on: moduleView!)
    } catch let err {
      result(FlutterError(code: "30007", message: err.localizedDescription, details: nil))
    }
  }

  public func setType(type: String, result: @escaping FlutterResult) {
    module.setType(type: type)
    result(nil)
  }

  public func upload(result: @escaping FlutterResult) {
    module.upload { isSuccess in
      result(isSuccess)
    }
  }
  
  public func setVideoRecording(enabled: Bool, result: @escaping FlutterResult) {
    module.setVideoRecording(enabled: enabled)
    result(nil)
  }
}

File: example/lib/screens/pose_estimation.dart
import 'package:flutter/material.dart';
import 'package:flutter_amanisdk_v2/amani_sdk.dart';
import 'package:flutter_amanisdk_v2/common/models/android/pose_estimation_settings.dart';
import 'package:flutter_amanisdk_v2/common/models/ios/pose_estimation_settings.dart';
import 'package:flutter_amanisdk_v2_example/screens/confim.dart';

class PoseEstimationScreen extends StatefulWidget {
  const PoseEstimationScreen({Key? key}) : super(key: key);

  @override
  State<PoseEstimationScreen> createState() => _PoseEstimationScreenState();
}

class _PoseEstimationScreenState extends State<PoseEstimationScreen> {
  final _amaniPoseEstimation = AmaniSDK().getPoseEstimation();

  static const routeName = "/pose-estimation";

  Future<void> initSDK() async {
    await _amaniPoseEstimation.setType("XXX_SE_0");
  }

  @override
  void initState() {
    super.initState();
    initSDK();
  }

  // Configure the pose estimation
  // Configure the AutoSelfie
  final AndroidPoseEstimationSettings _androidPoseEstimationSettings =
      AndroidPoseEstimationSettings(
    poseCount: 3,
    animationDuration: 500,
    faceNotInside: "Your face is not inside the area",
    faceNotStraight: "Your face is not straight",
    keepStraight: "Please hold stable",
    alertTitle: "Verification Failed",
    alertDescription: "Failed 1",
    alertTryAgain: "Try again",
    mainGuideVisibility: true,
    secondaryGuideVisibility: true,
  );

  final IOSPoseEstimationSettings _iosPoseEstimationSettings =
      IOSPoseEstimationSettings(
    faceIsOk: "Please hold stable",
    notInArea: "Please align your face to the area",
    faceTooSmall: "Your face is in too far",
    faceTooBig: "Your face is in too close",
    completed: "Verification Completed",
    turnRight: "→",
    turnLeft: "←",
    turnUp: "↑",
    turnDown: "↓",
    lookStraight: "Look straight",
    errorMessage:
        "Please complete the steps while your face is aligned to the area",
    tryAgain: "Try again",
    errorTitle: "Verification Failure",
    confirm: "Confirm",
    next: "Next",
    holdPhoneVertically: "Please hold the phone straight",
    informationScreenDesc1:
        "To start verification, align your face with the area",
    informationScreenDesc2: "",
    informationScreenTitle: "Selfie Verification Instructions",
    wrongPose: "Your face must be straight",
    descriptionHeader:
        "Please make sure you are doing the correct pose and your face is aligned with the area",
    appBackgroundColor: "000000",
    appFontColor: "ffffff",
    primaryButtonBackgroundColor: "ffffff",
    primaryButtonTextColor: "000000",
    ovalBorderColor: "ffffff",
    ovalBorderSuccessColor: "00ff00",
    poseCount: "3",
    mainGuideVisibility: "true",
    secondaryGuideVisibility: "true",
    buttonRadious: "10",
    manualCropTimeout: 30,
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Colors.deepPurple,
        title: const Text("Pose estimation"),
      ),
      body: Center(
        child: OutlinedButton(
          onPressed: () {
            _amaniPoseEstimation
                .start(
                    androidSettings: _androidPoseEstimationSettings,
                    iosSettings: _iosPoseEstimationSettings)
                .then((imageData) {
              Navigator.pushNamed(context, ConfirmScreen.routeName,
                  arguments: ConfirmArguments(
                      source: "poseEstimation", imageData: imageData));
            }).catchError((err) {
              throw Exception(err);
            });
          },
          child: const Text('Start Pose Estimation'),
        ),
      ),
    );
  }
}
